Encryption.md

```diff
@@ -0,0 +1,221 @@
+write/read throughput. The second Parquet algorithm encrypts the data content (pages) 
+with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
+the bulk of the data faster, while still verifying the metadata integrity and making sure 
+the file had not been replaced with a wrong version. However, tampering with the page data 
+might go unnoticed. 
+
+
+### AES_GCM_V1
+All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
+written after each ciphertext. The IVs (12 bytes) are either written before each ciphertext, 
+or split into two parts: a fixed part (n bytes) written in the `AesGcmV1` structure 
+(`iv_prefix` field), and a variable part (12-n bytes: e.g a counter) written before 
+each ciphertext.
+
+### AES_GCM_CTR_V1
+Thrift modules are encrypted with the AES-GCM cipher, as described above. The optional iv_prefix
+is stored in `AesGcmCtrV1` structure as a `gcm_iv_prefix` field.
+The pages are encrypted with AES-CTR, where the IVs (16 bytes) are either written before each 
+ciphertext, or split into two parts: a fixed part (n bytes) written in the `AesGcmCtrV1` structure 
+(`ctr_iv_prefix` field), and a variable part (16-n bytes: e.g. a counter) written before 
+each ciphertext.
+
+The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
+be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
+length of `aad_metadata` is 512 bytes.
+
+Parquet-mr/-cpp *writer* implementations use the RBG-based IV construction as defined in the NIST 
+SP 800-38D document for the GCM ciphers (section 8.2.2), and therefore don't set the iv_prefix fields. 
+However, the parquet-mr/-cpp *reader* implementations can process files with iv_prefix values, if 
+set by other writers that use the deterministic IV construction, defined in the section 8.2.1 in the 
+same document.
+
+
+## File Format
+
+The encrypted Parquet files have a different extension - “.parquet.encrypted”.
+
+The encryption is flexible - each column and the footer can be encrypted with the same key, 
+with a different key, or not encrypted at all.
+
+The metadata structures (`PageHeader`, `ColumnIndex`, `OffsetIndex`; and sometimes `FileMetaData` and 
+`ColumnMetaData`, see below) are encrypted after Thrift serialization. For each structure, 
+the encryption buffer is comprised of an IV, ciphertext and tag, as described in the 
+Algorithms section. The length of the encryption buffer (a 4-byte little endian) is 
+written to the output stream, followed by the buffer itself.
+
+The column pages (data and dictionary) are encrypted after compression. For each page, 
+the encryption buffer is comprised of an IV, ciphertext and (in case of AES_GCM_V1) of a 
+tag, as described in the Algorithms section. Only the buffer is written to the output 
+stream - not need to write the length of the buffer, since the length is kept in the page 
```

Comment:
Ah, I see. I was looking at DataPageHeader instead of PageHeader.

Speaking of compressed sizes: I think you should update the comments in the Thrift file to clarify that these sizes are about data that is compressed and potentially encrypted. (Also, is there any size field that is after compression but before encryption? Probably not.)

Reply:
This is indeed the case, modular encryption doesn't require additional fields for the size. I'll update the comments (and the doc) accordingly.